### DISCUSSION TITLE
How transform the extract images?

### DISCUSSION THREAD

**Discussion by hbh112233abc** [pymupdf/PyMuPDF#1896] [Looking for help] [answered]
I found some pdf page `transformation_matix`=`Matrix(1.0, 0.0, 0.0, -1.0, 0.0, 841.8900146484375)`, the page images extract output will flip like:
![image](https://user-images.githubusercontent.com/7788164/186361094-1330691e-2428-4a8e-9de9-ff9b920a3b2b.png)

my code:
```
image_no = 0
result = []
for page in self.doc.pages():
    page_images = page.get_images()
    breakpoint()
    for image in page_images:
        xref = image[0]
        width, height = image[2], image[3]
        if min(width, height) < dim_limit:
            continue
        img = self.doc.extract_image(xref)
        img_data = img["image"]
        pix = fitz.Pixmap(img["image"])
        if (len(img_data) / len(pix.samples)) < rel_size:
            continue
        if len(img_data) <= abs_size:
            continue
        image_no += 1
        if image_prefix:
            image_name = f"{image_prefix}-{image_no}.{img['ext']}"
        else:
            image_name = f"{image_no}.{img['ext']}"
        image_path = os.path.join(image_save_dir, image_name)
        with open(image_path, "wb") as f:
            f.write(img_data)
        result.append(image_path)
return result
```
I try use `Pixma.warp` function but not effect.

so how transform the image to normal style?
I hope extract image like:
![image](https://user-images.githubusercontent.com/7788164/186362612-fe902cc0-ba7a-45cd-acd1-15c41c1c62cc.png)

**Comment by JorjMcKie** [comment]
When the image is displayed in a bbox on a page, it must always be transformed.
At least for fitting into the bbox, but sometimes also things light rotation or left-right flipping are required.
What actually happened to the image, is contained in the special transformation matrix used for this adjustment. Get it like this:
```python
rects = page.get_image_rects(xref, transform=True)
# normally is a list with 1 item (multiple items if same image is displayed multiple times on the page)
rect, matrix = rects[0]
```
The transformation matrix fullfills the following equation - see docu details [here](https://pymupdf.readthedocs.io/en/latest/app3.html#imagetransformation):
```python
imgrect = fitz.Rect(0, 0, width, height)  # the original rectangle of the image
shrink = fitz.Matrix(1/width, 0, 0, 1/height, 0, 0)  # matrix converting image rect to unit rect
bbox = imgrect * shrink * transform  # this is how bbox on page is computed
```
So ``imgrect = bbox * ~transform * ~shrink``.

**Reply by hbh112233abc** [reply]
so how output the image not flipping? 
Could you help me, give some demo code? handle the image `pixmap`?

**Comment by JorjMcKie** [comment, marked as answer]
You can inspect the transformation matrix properties and compute the respective inverted matrix.
Then load the image in a pixmap and transform it with that matrix.
[matrix_property.zip](https://github.com/pymupdf/PyMuPDF/files/9439705/matrix_property.zip)

Depending on the return code of the function `matprop` build that inverted matrix. So if e.g. the return code is `(3, "rot 90")` take `~fitz.Matrix(90)`.

**Reply by hbh112233abc** [reply]
Thanks, use `matrix_property.py` I found the image matprop is `(2,'up-down')`, which function of `pixmap` can transform the image pixmap??

**Reply by hbh112233abc** [reply]
```
import fitz
from icecream import ic
from matrix_property import matprop

file = 'd:/py/base/test/投标文件.pdf'
pdf = fitz.open(file)
page = pdf[0]

img = page.get_images()[0]
ic(img)
xref = img[0]
img_name = img[7]
img_width = img[2]
img_height = img[3]
imgrect = fitz.Rect(0,0,img_width,img_height)
ic(imgrect)
shrink = fitz.Matrix(1/img_width,0,0,1/img_height,0,0)
ic(shrink)
bbox, transform = page.get_image_bbox(img_name, transform=True)
ic(bbox)
ic(transform)
ic(imgrect*shrink*transform)

img = pdf.extract_image(xref)
img_data = img["image"]
pix = fitz.Pixmap(img["image"])

img_pix = fitz.Pixmap(pdf,xref)
ic(img_pix)

ic(matprop(transform))
```
output
```
ic| img: (8, 0, 1001, 1001, 8, 'DeviceRGB', '', 'IM8', 'DCTDecode')
ic| imgrect: Rect(0.0, 0.0, 1001.0, 1001.0)
ic| shrink: Matrix(0.000999000999000999, 0.0, 0.0, 0.000999000999000999, 0.0, 0.0)
ic| bbox: Rect(166.9199981689453, 483.84002685546875, 287.0400085449219, 603.9600219726562)
ic| transform: Matrix(120.12000274658203, 0.0, -0.0, -120.12000274658203, 166.9199981689453, 603.9600219726562)
ic| imgrect*shrink*transform: Rect(166.9199981689453, 483.84002685546875, 287.0400085449219, 603.9600219726562)
ic| img_pix: Pixmap(DeviceRGB, IRect(0, 0, 1001, 1001), 0)
ic| matprop(transform): (2, 'up-down')
```
the image pixmap can load by image stream or `fitz.Pixmap(doc,xref)`, but I can not found any function to transform the image pixmap!

**Reply by JorjMcKie** [reply]
```python
flip = fitz.Matrix(1,0,0,-1,0,0)  # flips up-down, see documentation
img = fitz.open(<...>)  # image filename or bytes
pix = img[0].get_pixmap(matrix=flip)
pix.save("...") # or use pil_save to store as JPEG
```

**Reply by hbh112233abc** [reply]
Thank you very much! I get the images with my hope style 👍 
I suggest add some function in `Pixmap` to transform pixmap with a `Matrix`, I think it's useful!

**Reply by JorjMcKie** [reply]
> I suggest add some function in Pixmap to transform pixmap with a Matrix

The Pillow package already has this type of functions. So adding similar functionality would mean re-inventing the wheel - although it could be done with a somewhat better performance probably ... 😉.
Convert the pixmap to a PIL.Image as described in the documentation and then apply the resp. flip / rotate method as described in Pillow doc.

**Reply by hbh112233abc** [reply]
thanks again!
